src/java_bytecode/generate_java_generic_type.cpp

```diff
@@ -321,32 +272,56 @@ irep_idt generate_java_generic_typet::build_generic_name(
       existing_generic_type.generic_type_arguments().size() >=
         implicitly_generic_original_class.implicit_generic_types().size(),
       "All implicit generic types must be concretised");
-    auto implicit_generic_type_p =
-      implicitly_generic_original_class.implicit_generic_types().begin();
-    const auto &implicit_generic_types_end =
-      implicitly_generic_original_class.implicit_generic_types().end();
+
+    std::string previous_outer_class_name;
     std::string current_outer_class_name;
+    generic_arguments_name_buildert build_generic_arguments(
+      argument_name_printer);
 
-    while(implicit_generic_type_p != implicit_generic_types_end)
+    // add generic arguments to each generic outer class
+    for(const auto &implicit_generic_type :
+        implicitly_generic_original_class.implicit_generic_types())
     {
+      previous_outer_class_name = current_outer_class_name;
       current_outer_class_name =
-        (*implicit_generic_type_p).get_parameter_class_name();
-      generic_name_buffer << current_outer_class_name;
-      generic_name_buffer << build_generic_name_for_class_arguments(
-        generic_argument_p,
-        implicit_generic_type_p,
-        implicit_generic_types_end,
-        current_outer_class_name);
+        implicit_generic_type.get_parameter_class_name();
+
+      // if the class names do not match, it means that the next generic
+      // outer class is being handled
+      if(current_outer_class_name != previous_outer_class_name)
+      {
+        // add the arguments of the previous generic outer class to the
+        // buffer and reset the builder
+        generic_name_buffer << build_generic_arguments.finalize();
+        generic_arguments_name_buildert build_generic_arguments(
+          argument_name_printer);
+
+        // precondition to check that the implicit generic types are ordered
+        // from the outermost outer class inwards
+        PRECONDITION(
+          has_prefix(current_outer_class_name, previous_outer_class_name));
+
+        // add the remaining part of the current generic outer class name to
+        // the buffer
+        generic_name_buffer << current_outer_class_name.substr(
```

Comment:
Consider adding an examples of what `original_class_name`, `current_class_name` will be so people can see how this is working.

Reply:
E.g 
```C++
// current_outer_class = A.B.C
// previous_outer_class = A
// Then substr of current, starting at the length of previous is .B.C
```